radicle-node/src/service/reactor.rs

```diff
@@ -31,12 +31,42 @@ pub enum Io {
 pub struct Fetch {
     /// Repo to fetch.
     pub rid: Id,
-    /// Namespaces to fetch.
-    pub namespaces: Namespaces,
+    /// Indicates whether the fetch request was initiated or is a response.
+    pub direction: FetchDirection,
     /// Remote peer we are interacting with.
     pub remote: NodeId,
-    /// Indicates whether the fetch request was initiated by us.
-    pub initiated: bool,
+}
+
+impl Fetch {
+    pub fn initiated(&self) -> bool {
+        self.direction.initiated()
+    }
+
+    pub fn namespaces(&self) -> Option<&Namespaces> {
+        match &self.direction {
+            FetchDirection::Receive { namespaces } => Some(namespaces),
+            FetchDirection::Upload => None,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum FetchDirection {
+    /// Client is fetching in order to receive the specified refspecs
+    /// in [`Namespaces`].
+    Receive {
+        /// Namespaces to fetch.
+        namespaces: Namespaces,
+    },
+    /// Server is responding to a fetch request by uploading the
+    /// specified refspecs.
+    Upload,
+}
```

Comment:
I think we can do a bit better here in terms of the variant names; especially I think `Receive` doesn't communicate enough (it might sound like you are receiving the fetch request).

Some proposals:
`Initiator / Responder` (this terminology is from Noise and I quite like it)
`Download / Upload`
`Fetch / Upload` (git-influenced)

Reply:
On this topic, I'd also try to have the `is_` functions and the one returning the namespaces match the naming of the variants better.

Reply:
Ah gotcha.